### PR TITLE
W-13830961 configuring features gr

### PR DESCRIPTION
**Pull request by glenn-rodgers-sf**
# Writer's Quality Checklist

Before merging your PR, did you:

- [ ] Run spell checker
- [ ] Run link checker to check for broken xrefs
- [ ] Check for orphan files
- [ ] Perform a local build and do a final visual check of your content, including checking for:
  - Broken images
  - Dead links
  - Correct rendering of partials if they are used in your content
  - Formatting issues, such as:
    - Misnumbered ordered lists (steps) or incorrectly nested unordered lists
    - Messed up tables
    - Proper indentation
    - Correct header levels
- [ ] Receive final review and signoff from:
  - Technical SME
  - Product Manager
  - Editor or peer reviewer
  - Reporter, if this content is in response to a reported issue (internal or external feedback)
- [ ] If applicable, verify that the software actually got released

### CODE REVIEW

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 1]
```diff
@@ -1,5 +1,586 @@
...
+Currently, the policy performs some simple logging. For each incoming request, the policy logs the header `token` at the `info` log level.
+
+This is complete by the following elements:
+
+* `use pdk::api::hl::*;`: Imports all the necessary components of the PDK into the `lib.rs` source code.
+* `#[entrypoint]` function: This function executes when the policy is applied and calls the `request_filter` function`. Variables defined inside this function are avaliable while the policy is applied.
++
+This `#[entrypoint]` function receives the following parameters and passes them to the `request_filter` function`:
++
+** `Launcher`: Provides information about incoming request, such as headers and request type.
```
Comment (JaatadiaMulesoft) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]: the launcher sets the filter functions as the functions to be used.

**Review by JaatadiaMulesoft — CHANGES_REQUESTED** [review]

**Review by JaatadiaMulesoft — APPROVED** [review]

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 2]
```diff
@@ -1,5 +1,589 @@
...
+* xref:policies-pdk-create-project.adoc[Create a New Project]
+* xref:policies-create-schema-definition.adoc[Define a Schema Definition]
++
+NOTE: You may need to add or redefine configuration parameters while implementing your policy. 
+
+* Review <<policy-template>> before adding additional features.
+
+== Policy Template
+
+PDK provides the initial `src/lib.rs` file in the policy project as a template to begin impletmenting your policy:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
PDK provides the initial `src/lib.rs` file in the policy project as a template to begin implementing your policy:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 3]
```diff
@@ -1,5 +1,589 @@
...
+
+* `use pdk::api::hl::*;`: Imports all the necessary components of the PDK into the `lib.rs` source code.
+* `#[entrypoint]` function: Executes when the policy is applied and calls the `request_filter` function. Variables defined inside this function are avaliable while the policy is applied.
++
+The `#[entrypoint]` function receives the following parameters:
++
+** `Launcher`: Sets the filter functions.
+** `Configuration`: Provides the configuration parameters defined in xref:policies-pdk-create-schema-definition.adoc[]. 
+
+* `request_filter`: Executes once per every request sent to the API instance to which the policy is applied. This function is the implementation of the filtering performed by the policy. Variables defined inside this function available for the duration of the request.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
* `request_filter`: Executes once per every request sent to the API instance to which the policy is applied. This function is the implementation of the filtering performed by the policy. Variables defined inside this function are available for the duration of the request.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 4]
```diff
@@ -1,5 +1,589 @@
...
++
+The `request_filter` is an example of a wrapped function
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
The `request_filter` is an example of a wrapped function.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 5]
```diff
@@ -1,5 +1,589 @@
...
++
+The `#[entrypoint]` function executes wrapped functions:
++
+[source,Rust]
+----
+let filter = on_request(|rs| request_filter(rs, &config));
+launcher.launch(filter).await?;
+----
+
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc:80]: Is it possible to add a sub-section here? Otherwise, the example is too long to follow. Something like: 
=== Processing the Outgoing Response
=== Using an on_response Wrapper
Not sure if these titles are accurate, but something like this would help understand that this is some additional information to the first policy template.

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 6]
```diff
@@ -1,5 +1,589 @@
...
+    let config = serde_json::from_slice(&bytes)?;
+    let filter = on_response(|rs| response_filter(rs, &config));
+    launcher.launch(filter).await?;
+    Ok(())
+}
+----
+
+NOTE: The `request_filter` function is now the `response_filter` function.
+
+To filter both requests and response, use bother wrappers:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
To filter both requests and responses, use bother wrappers:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 7]
```diff
@@ -1,5 +1,589 @@
...
+async fn configure(launcher: Launcher, Configuration(bytes): Configuration) -> Result<()> {
+    let config = serde_json::from_slice(&bytes)?;
+    let filter = on_request(|rs| request_filter(rs, &config))
+        .on_response(|rs| response_filter(rs, &config));
+    launcher.launch(filter).await?;
+    Ok(())
+}
+----
+
+The `on_response` or `on_request` wrapper defines when the request is executed.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
The `on_response` or `on_request` wrapper defines when the request executes. 
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 8]
```diff
@@ -1,5 +1,589 @@
...
+    let filter = on_request(|rs| request_filter(rs, &config))
+        .on_response(|rs| response_filter(rs, &config));
+    launcher.launch(filter).await?;
+    Ok(())
+}
+----
+
+The `on_response` or `on_request` wrapper defines when the request is executed. 
+
+You can define functions other than the provided `requests_filter` and `response_filter` and define them from the `#[entrypoint]` function. The `requests_filter` and `response_filter`` are only examples.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
You can define functions other than the provided `requests_filter` and `response_filter` from the `#[entrypoint]` function. The `requests_filter` and `response_filter` are only examples.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 9]
```diff
@@ -1,5 +1,589 @@
...
+PDK provides a logging mechanism that generates a log message enriched with the API instance ID, policy ID, and request ID.
+
+Insert custom logs with the following macros by using the `logger::` package:
+
+* `logger::debug!`
+* `logger::info!`
+* `logger::warn!`
+* `logger::error!`
+
+The macros behave the same as the https://doc.rust-lang.org/std/macro.format.html[Rust std::format! macro].  The first parameter must be a format string literal. Use `{}` in the literal to pass parameters, for example:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
The macros behave the same as the https://doc.rust-lang.org/std/macro.format.html[Rust std::format! macro^]. The first parameter must be a format string literal. Use `{}` in the literal to pass parameters, for example:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 10]
```diff
@@ -1,5 +1,589 @@
...
+All examples appear in the Flex Gateway logs in the following format:
+----
+[flex-gateway-envoy][<log-level>] wasm log <policy-name>.<api-instance-name> main: [policy: <policy-name>][api: <api-instance-name>][req: <request-id> ] Hello there!
+----
+
+For more information about viewing Flex Gateway logs, see xref:flex-gateway-monitor.adoc[].
+
+== Read and Write Headers
+
+To access the headers in both the request and the response, transform the `RequestState` or `ResponseState` to a header state by calling the method `into_headers_state` and awaiting it. After calling either method, access and manipulate the headers by calling the functions of the `HeadersHandler` trait.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc:159]: "and awaiting it." 
Not sure what this means, do you wait for the method to complete? Do users have to set some kind of listener that waits until the method finishes? Can't we just remove this last part and end the sentence with "by calling the method `into_headers_state`"? Please confirm.

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 11]
```diff
@@ -1,5 +1,589 @@
...
+    let old_headers = headers_state.header("request-header").unwrap_or_default();
+
+    let new_value = vec![("response-header1", "--replaced--"), ("response-header2", "--replaced--")];
+    logger::info!("Old request header value: {old_headers:?}, New value: {new_value:?}");
+    headers_state.set_headers(new_value);
+}
+
+----
+
+Envoy handles the method, scheme, path, authority, and status codes as headers. Modify the headers by accessing the request's `:method`, `:scheme`, `:path`, `:authority`, and `:status`. Additionally PDK provides the `RequestHeadersHandler` and `ResponseHeadersHandler` traits for ease of access:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Envoy handles the method, scheme, path, authority, and status codes as headers. Modify the headers by accessing the request's `:method`, `:scheme`, `:path`, `:authority`, and `:status`. PDK provides the `RequestHeadersHandler` and `ResponseHeadersHandler` traits for ease of access:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 12]
```diff
@@ -1,5 +1,589 @@
...
+}
+
+pub trait ResponseHeadersHandler: HeadersHandler {
+    fn status_code(&self) -> u32;
+}
+----
+
+== Read the Request Body
+
+To access the body in both the request and the response, transform the `RequestState` or `ResponseState` to a body state by calling the method `into_body_state` and awaiting it. If the original state was already transformed into a header state, transform the state into a body state by calling the same function, for example:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc:215]: Same comment about "awaiting it"

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 13]
```diff
@@ -1,5 +1,589 @@
...
+After calling `into_body_state`, access and manipulate the headers by calling the functions of the `BodyHandler` trait.
+
+[source,Rust]
+----
+pub trait BodyHandler {
+    fn body(&self) -> Vec<u8>;
+}
+----
+
+Envoy uses the same buffer to share data from the headers and the body to the policy. So, at a given moment, the policy can only access the headers or the body. If the policy must read both:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Because Envoy uses the same buffer to share data from the headers and the body, the policy cannot access the headers and the body at the same time. If the policy must read both:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 14]
```diff
@@ -1,5 +1,589 @@
...
+
+    launcher
+        .launch(filter)
+        .await?;
+
+    Ok(())
+}
+----
+
+If your `on_request` or `on_response` wrapped functions only receives the predefined parameters, it is not necessary to include the lambda in the wrapper. For example, if the signature of the function is:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
If your `on_request` or `on_response` wrapped functions only receive the predefined parameters, the lambda is not required in the wrapper. For example, if the signature of the function is:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 15]
```diff
@@ -1,5 +1,589 @@
...
+
+Instead of a function definition of:
+
+[source,Rust]
+----
+let filter = on_request(|request_state| request_filter(request_state))
+    .on_response(|response_state, request_data| response_filter(response_state, request_data));
+----
+
+You can use the following functions definition:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
You can use the following function definition:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 16]
```diff
@@ -1,5 +1,589 @@
...
+
+[source,Rust]
+----
+let filter = on_request(request_filter)
+        .on_response(response_filter);
+----
+
+== Perform an HTTP Call
+
+If your policy must interact with external services, the policy must use HTTP calls. You can preform HTTP calls in the request and response filters and in the function. To perform a the HTTP call, inject the `HttpClient` as follows:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
To interact with external services, the policy must use HTTP calls. You can perform HTTP calls in the request and response filters and in the function. To perform an HTTP call, inject the `HttpClient` as follows:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 17]
```diff
@@ -1,5 +1,589 @@
...
+[source,Rust]
+----
+​​async fn request_filter(state: RequestState, conf: &Config, client: HttpClient) {
+ ...
+}
+----
+
+You can only make HTTP calls to services defined in the Flex Gateway configuration files. For more information about defining services, see xref:flex-local-configuration-reference-guide.adoc#service[Services].
+
+For both Connected and Local mode, define services using `Service` resource. If the service is not defined when the request is made, the API request fails.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
For both Connected and Local modes, define services using the `Service` resource. If the service is not defined when the request is made, the API request fails.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 18]
```diff
@@ -1,5 +1,589 @@
...
+----
+fn is_null(&self) -> bool;
+fn as_bool(&self) -> Option<bool>;
+fn as_f64(&self) -> Option<f64>;
+fn as_str(&self) -> Option<&str>;
+fn as_slice(&self) -> Option<&[Value]>;
+fn as_object(&self) -> Option<&Object>;
+----
+
+The functions return an `Option` containing the value if it was properly casted or `None` if the value didn't belong to the casted type. For example, the following code snippet, tests what type the value resolves as:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
The functions return an `Option` containing the value if it was properly cast or `None` if the value didn't belong to the cast type. For example, the following code snippet tests what type the value resolves as:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 19]
```diff
@@ -1,5 +1,589 @@
...
+[source,Rust]
+----
+async fn request_filter(_request_state: RequestState) -> Response {
+    Response::new(401)
+        .with_headers(vec![("WWW-Authenticate".to_string(), "Bearer realm=\"oauth2\"".to_string())])
+        .with_body(r#"{ "error": "token was not present"}"#)
+}
+----
+
+This function is not very useful as it intercepts the requests no matter what. You can block or allow requests to reach the upstream service using a `Flow`. A `Flow` in an `enum` value with two possible values:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]: If this function is not useful, we should present it as an alternative and show the second option (You can block or allow requests to reach the upstream service using a `Flow`) as the first one. 

Also, I'd rephrase the introduction to this function to something like: "In case you need to intercept and stop the request under any circumstances, return a `Response` object from your `on_request` wrapped functions"

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 20]
```diff
@@ -1,5 +1,589 @@
...
+        Flow::Break(Response::new(401)
+        .with_headers(vec![("WWW-Authenticate".to_string(), "Bearer realm=\"oauth2\"".to_string())])
+        .with_body(r#"{ "error": "token was not present"}"#))
+    }
+}
+----
+
+[IMPORTANT]
+====
+Due to the streaming nature of `proxy wasm`, it is possible for an early request to partially reach the upstream service while awaiting the `headers_state.into_body_state` function. The only way to avoid this by not awaiting the body. However, that is not possible for all policies. You should configure your upstream to ignore partial requests if the policy must read the body.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Due to the streaming nature of `proxy wasm`, it is possible for an early request to partially reach the upstream service while awaiting the `headers_state.into_body_state` function. The only way to avoid this is by not awaiting the body. However, that is not possible for all policies. Configure your upstream to ignore partial requests if the policy must read the body.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 21]
```diff
@@ -1,5 +1,589 @@
...
+    let filter = on_request(request_filter)
+        .on_response(response_filter);
+
+    launcher
+        .launch(filter)
+        .await
+}
+----
+
+Returning a `RequestData` makes it impossible for a response to reach a client early. This is why the `Flow::Continue` enum has a template parameter. Implement this by adding a `on_response` wrapped function to the `Flow` example used in <<stop-request-execution>>:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Returning a `RequestData` prevents a response from reaching a client early. This is why the `Flow::Continue` enum has a template parameter. Implement this by adding an `on_response` wrapped function to the `Flow` example used in <<stop-request-execution>>:
```
"This is why the `Flow::Continue` enum has a template parameter" It is not clear for me what is the template parameter mentioned here for, can you please update this sentence to something that explains why this is important? For example: 
- "Use the template parameter in the `Flow::Continue` enum to...", or 
- "The `Flow::Continue` enum has a template parameter that..."

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 22]
```diff
@@ -1,5 +1,589 @@
...
+
+async fn response_filter(state: ResponseState, RequestData(path): RequestData<String>) {
+    let state = state.into_headers_state().await;
+    logger::info!("Path: {path}, Status: {}", state.status_code());
+}
+----
+
+== Share Data Between Workers and Caching
+
+To increase the capability of handling requests, envoy spawns many workers for the policy. A single worker handles all the stages of a specific request including outgoing connections and the response flow. Due to this, you can track an internal status for a specific request and create sequential execution flow even though the underlying implementation is event driven.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Envoy spawns many workers for the policy to increase the capability of handling requests. A single worker handles all the stages of a specific request, including outgoing connections and the response flow, enabling you to track an internal status for a specific request and create a sequential execution flow even though the underlying implementation is event-driven. 
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 23]
```diff
@@ -1,5 +1,589 @@
...
+}
+----
+
+== Share Data Between Workers and Caching
+
+To increase the capability of handling requests, envoy spawns many workers for the policy. A single worker handles all the stages of a specific request including outgoing connections and the response flow. Due to this, you can track an internal status for a specific request and create sequential execution flow even though the underlying implementation is event driven. 
+
+Each worker processes a single request at a given time. To share a global state in a specific worker, use a `RefCell` without any risk of concurrent modification.
+
+Communicating a shared global state to all workers is more complex. Achieve this is by using the provided `Cache` mechanism.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Use the provided `Cache` mechanism to communicate a shared global state to all workers.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 24]
```diff
@@ -1,5 +1,589 @@
...
+
+After injecting the cache builder, provide an ID for the cache and the maximum number of elements the cache can support. 
+
+For example a cache with id “calculator” can hold ten elements:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
In the following example, a cache with ID “calculator” can hold ten elements:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 25]
```diff
@@ -1,5 +1,589 @@
...
+----
+pub trait Cache {
+    fn save(&self, key: &str, value: Vec<u8>) -> Result<(), CacheError>;
+    fn get(&self, key: &str) -> Option<Vec<u8>>;
+    fn delete(&self, key: &str) -> Option<Vec<u8>>;
+    fn purge(&self);
+}
+----
+
+IMPORTANT: Because another work may write to the cache concurrently, there is no guarantee that the value overwritten when saving the cache is the same as when it was retrieved.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
IMPORTANT: Because different works write to the cache concurrently, there is no guarantee that the value overwritten when saving the cache is the same as when it was retrieved.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 26]
```diff
@@ -1,5 +1,589 @@
...
+
+== Use Third-Party Libraries
+ 
+Proxy Wasm defines a low level binary application interface (abi), that limits the system calls a library can make. All third-party libraries included in your policy must be compatible with the `wasm32-wasi` Rust compilation target.
+
+To define a library, such as `serde_urlencoded` has already been, add it to the depenpedency list in `cargo.toml`:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
To define a library such as `serde_urlencoded`, add it to the dependency list in `cargo.toml`:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 27]
```diff
@@ -1,5 +1,589 @@
...
+
+[source,toml]
+----
+[dependencies]
+...
+serde_urlencoded = "0.7.0"
+----
+
+This enables you to use functions in `lib.rs` such as:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Defining libraries enables you to use functions in `lib.rs` such as:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 28]
```diff
@@ -1,5 +1,589 @@
...
+
+[source,Rust]
+----
+serde_urlencoded::to_string([("token", "myToken")])
+----
+
+In general, if a library interacts with an external service, for example a database, or performs a system call, like reading a file, it is probable that it isn't compatible with the `wasm32-wasi` target.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
Libraries that interact with an external service like a database, or perform a system call like reading a file, are usually not compatible with the `wasm32-wasi` target. 
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 29]
```diff
@@ -1,5 +1,589 @@
...
+This enables you to use functions in `lib.rs` such as:
+
+[source,Rust]
+----
+serde_urlencoded::to_string([("token", "myToken")])
+----
+
+In general, if a library interacts with an external service, for example a database, or performs a system call, like reading a file, it is probable that it isn't compatible with the `wasm32-wasi` target. 
+
+If your policy to must interact with external services, achieve this is using Flex Gateway's exposed HttpClient. For more information about performing an HTTP call, see <<performing-an-http-call>>.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
If your policy interacts with external services, use the Flex Gateway exposed HttpClient. For more information about performing an HTTP call, see <<performing-an-http-call>>.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 30]
```diff
@@ -1,5 +1,589 @@
...
+
+Even though some libraries may compile properly to the wasm32-wasi target, they still might not work properly when deployed to Flex Gateway. Example errors of this are:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
It is possible that some libraries compile properly to the `wasm32-wasi` target but don't work properly when deployed to Flex Gateway. Example errors of this are:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc [hunk 31]
```diff
@@ -1,5 +1,589 @@
...
+
+* `Failed to load Wasm module due to a missing import: ...`
+* `Wasm VM failed Failed to initialize Wasm code`
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc]:
```suggestion
* `Wasm VM failed to initialize Wasm code`
```
This might be a typo on the error itself, but since it is not an error code maybe we can fix the message here.

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc (hunk 10 above)
Comment (glenn-rodgers-sf) [reply gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc:159]: I'm not exactly sure how it works but since Rust is a multi-thread language there's a .operator calles "await" that signifies dependencies. This is the function used in the code below "request_state.into_headers_state().await".

Sometimes you need to await functions but sometimes you do not.

File: gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc (hunk 12 above)
Comment (glenn-rodgers-sf) [reply gateway/1.5/modules/ROOT/pages/policies-pdk-configure-features.adoc:215]: Same as above

**Review by Cristian-Venticinque — APPROVED** [review]
